### ISSUE TITLE
<any>: Relying on typeid causes incompatabilities between MSVC and Clang

### ISSUE THREAD

**Issue by joshuamaiche** [microsoft/STL#4349]
# Overview
`<any>` relies on `typeid`. When compiling with Clang with RTTI disabled, this used to cause compilation errors whenever `<any>` was included (https://github.com/microsoft/STL/issues/3114). This was fixed by skipping over `<any>` if `_HAS_STATIC_RTTI` was set to 0.

However, this now causes inconsistencies between compilers (with RTTI disabled).
- On compilers that use MSVC STL and lose `typeid` with RTTI off (Clang on Windows), `std::any` is not defined.
- On compilers that use MSVC STL and keep `typeid` with RTTI on (MSVC), `std::any` is defined.
- On compilers that use a different STL (GCC, Clang on Linux), `std::any` is defined.

Ideally, `<any>` would avoid depending on `typeid`, so there could be consistent behavior between compilers when disabling RTTI.

# Repro
```
// main.cpp

#include <any>

int main()
{
    std::any a = 0;
}
```

Successful builds:
`Windows>cl /GR- /std:c++17 main.cpp`
`Linux>g++ -std=c++17 -fno-rtti main.cpp`
`Linux>clang++ -std=c++17 -fno-rtti main.cpp`

Unsuccessful builds:
```
Windows>clang++ -std=c++17 -fno-rtti main.cpp
main.cpp:7:5: error: use of undeclared identifier 'std'
    std::any a = 0;
    ^
1 error generated.
```

# STL version
```
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.8.5
```

# Related bugs
https://github.com/nlohmann/json/issues/3900
https://github.com/llvm/llvm-project/issues/79677

**Comment by StephanTLavavej** [comment]
Thanks for reporting this issue. We talked about it at the weekly maintainer meeting, and we have no plans to change our implementation here. Not only is RTTI a part of the C++ Core Language (so we feel entirely justified in normally relying on it), but we also assume that we can rely on the MSVC compiler's behavior, and therefore we generally expect Clang to be source-compatible with such MSVC behavior. Here, we've gone substantially out of our way to cater to MSVC's non-Standard RTTI-disabled mode `/GR-`, and gone even further out of our way to implement and support an option to eradicate all usage of "static RTTI". If this remains insufficient for Clang targeting Windows, then we would recommend that it should more closely imitate MSVC's behavior under `/GR-`.

Not only are we locked into the current representation of `std::any` for the lifetime of the v19 binary-compatible release series, but even in a future vNext ABI-breaking release, altering `std::any`'s representation to handle RTTI-disabled scenarios for non-MSVC compilers falls under our documented [non-goals](https://github.com/microsoft/STL#non-goals) - we focus on Standard conformance and have very limited time/energy to handle non-Standard scenarios.
